### ISSUE TITLE
Simplify default expression (IDE0034) false positive causes the code fix to break behavior

### ISSUE THREAD

**Issue by HellBrick** [dotnet/roslyn#21638]
**Version Used**: 
VS 15.3.1

**Steps to Reproduce**:

```xml
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>netcoreapp2.0</TargetFramework>
		<LangVersion>latest</LangVersion>
	</PropertyGroup>
</Project>
```
```csharp
using System;

namespace DefaultLiteralCodeFixFalsePositive
{
	internal class Program
	{
		private static void Main( string[] args )
		{
			bool condition = args.Length > 0;
			int? value = Invoke( () => condition ? 42 : default( int? ) );
			Console.WriteLine( $"{condition} --> {value}" );
		}

		private static int? Invoke( Func<int?> factory ) => factory();
	}
}
``` 

**Expected Behavior**:
IDE0034 is **not** reported and the corresponding code fix is **not** suggested.

**Actual Behavior**:
IDE0034 is reported and the code fix changes the behavior of the false branch from returning `null` to returning `0`:

![The code fix suggestion](http://i.imgur.com/3RDchfz.png)
![The type being inferred incorrectly](http://i.imgur.com/vAbUpLj.png)

**Comment by HellBrick** [comment]
Oh, and here's another false positive:
```csharp
	public struct EmptyStruct
	{
		public static object BoxedInstance { get; } = default( EmptyStruct );
	}
```
![Boxing conversion false positive](http://i.imgur.com/5LR4QPY.png)

**Comment by Pilchie** [comment]
Are these cases already covered by https://github.com/dotnet/roslyn/issues/21581?

**Comment by CyrusNajmabadi** [comment]
Yes.  Both of hese cases were fixed by https://github.com/dotnet/roslyn/pull/20856

**Comment by HellBrick** [comment]
The second one definitely is, not sure about the first one: I don't see the `flag ? 42 : default( int? )` scenario mentioned either in that issue or in the tests added in the linked PR.

**Comment by CyrusNajmabadi** [comment]
changing ```default(int?)``` to ```default``` changes its type from ```int?``` to ```int```.  It is covered (and i have verified this).
